File: JavaScript/0383-赎金信.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
function canConstruct(ransomNote, magazine) {
  const rn = ransomNote.length,
    mn = magazine.length;
  if (rn > mn) {
    return false;
  }
  const hash = new Array(26).fill(0),
    base = 'a'.charCodeAt();

  for (let i = 0; i < mn; i++) {
    hash[magazine[i].charCodeAt() - base]++;
  }
  for (let i = 0; i < rn; i++) {
    if (--hash[ransomNote[i].charCodeAt() - base] < 0) {
      return false;
    }
  }
  return true;
}

File: Java/0383-赎金信.java
class Solution {
  public boolean canConstruct(String ransomNote, String magazine) {
    if (ransomNote.length() > magazine.length()) {
      return false;
    }
    int[] hash = new int[26];

    for (char c : magazine.toCharArray()) {
      hash[c - 'a']++;
    }

    for (char c : ransomNote.toCharArray()) {
      if (--hash[c - 'a'] < 0) {
        return false;
      }
    }
    return true;
  }
}
